Spain (Congreso): Add Twitter list scraper to sources

octopusinvitro opened the pull request:
Scraper is here: [https://morph.io/everypolitician-scrapers/twitter-spain-congreso-list](https://morph.io/everypolitician-scrapers/twitter-spain-congreso-list)

It looks like for some people it is not taking the Twitter avatar, even though I have gone there and there is indeed an image ¿?

A fewer amount don't even have a Twitter url.

Gonna investigate that when I finish going through `sources/reconciliation/twitter.html`

Closes everypolitician/everypolitician#471

tmtmtmtm commented:
@octopusinvitro you need to commit the reconciliation file you've generated, and then also the result of rebuilding with this source. Then we'll be able to look at the problematic cases.

octopusinvitro commented:
If I go to `morph/twitter.csv`, the links to the images are there, so it may be `twitter.html` the one not showing them.

These are some of the ones not showing a picture:

* https://twitter.com/tcunillera
* https://twitter.com/mariobedera
* https://twitter.com/TeresaJorda;TeresaJorda
* https://twitter.com/gmariscalanaya
* https://twitter.com/daniserrano2015 also no twitter info!
* https://twitter.com/mariajsusinos
* https://twitter.com/Mbonilladomingu
* https://twitter.com/carlos_corcuera
* https://twitter.com/Concepcioveray

Also I got two people with two Twitter ids:

**Consuelo Rumí:**
```
11d43b6c-82c9-4c5a-834d-bba5f37b6042 has these IDs:

        2909688201 RumiConsuelo NO
        384816005  ConsueloRumi YES
```
i.e.:
* ConsueloRumi: https://twitter.com/ConsueloRumi --> Deputy for Almería
* RumiConsuelo: https://twitter.com/RumiConsuelo --> Councillor for Almería 
I deleted the first one manually.

**Concepción (Concha) Gutiérrez:**
```
5bfa6927-a3a8-4952-9fc7-32693a7e0dfa has these IDs:

        1147635685 GutierrezConcha YES
        337595761  ConchaGutierrez YES
```
* ConchaGutierrez https://twitter.com/ConchaGutierrez --> ?
* GutierrezConcha https://twitter.com/GutierrezConcha --> Deputy 
In this case it was not clear so I left both.

octopusinvitro commented:
These two members:
* 307570052, Ana Pastor Julian, https://twitter.com/anapastorjulian
* 442111814, Rocío López González, https://twitter.com/rociolopez1905

Who need to get their Twitter handles updated with the ones shown above, as the ones we have stored don't exist anymore, because they changed their Twitter handle.

But I'm not sure exactly how to fix that.

tmtmtmtm commented:
@octopusinvitro this is revealing an interesting edge case with how the Reconciler works. You have two different source files sharing the same reconciliation file, which sort-of works, in that both of those sources share the same IDs, but the fact that it _does_ work is largely an accident, and it's not guaranteed to continue to work. 

I think this would be much better if you combined both of these sources into the same Morph scraper, and then just imported that as a single source.

tmtmtmtm commented on data/Spain/Congress/ep-popolo-v1.0.json:
```diff
@@ -55616,8 +55092,10 @@
+      [
+        "https://twitter.com/Congreso_Es/lists/diputados-2011-2016",
+        "https://twitter.com/Congreso_Es/lists/diputados-xi-legislatura"
+      ]
```
We’re currently blocked on https://github.com/everypolitician/everypolitician/issues/471 resolving this

octopusinvitro replied:
I just fixed it! :)
But I am on the bus to Bristol without internet on the laptop.
I will push it tomorrow as soon as I am in the office!

tmtmtmtm commented on rake_build/turn_csv_to_popolo.rb:
```diff
@@ -13,8 +13,7 @@
     @json[:meta] ||= {}
-    # TODO: allow for more than one source
-    @json[:meta][:sources] = instructions(:sources).map { |s| s[:source] }.compact.uniq
+    @json[:meta][:sources] = instructions(:sources).map { |s| s[:source] }.compact.uniq.flatten
```
:+1:

Though to be slightly neater you could do this with `instructions(:sources).flat_map { |s| s[:source }…` rather than a separate `flatten` step.

NB: You should also [close the issue from the commit](https://help.github.com/articles/closing-issues-via-commit-messages/).

octopusinvitro replied:
ooh, I didn't know about `flat_map`. Let me rebase-edit.

tmtmtmtm commented:
@octopusinvitro There seem to be a few here that you've missed. For example:

![screen shot 2016-07-25 at 14 40 16](https://cloud.githubusercontent.com/assets/57483/17103330/c2a0d512-5275-11e6-93db-7510a7ec4e06.png)

Pro Tip: if you temporarily change the `instructions.json` so that both the `incoming_field` and `existing_field` are "twitter", and regenerate the Reconciliation interface, it'll pick up all the outstanding people who match based on us already knowing their twitter handle…

octopusinvitro commented:
@tmtmtmtm More tribal knowledge for my wiki! yay :)

tmtmtmtm commented:
@octopusinvitro the commit message for https://github.com/everypolitician/everypolitician-data/pull/14239/commits/7294501bb7032705b623aba715ffcd594733b869 is a bit misleading.  "Regenerate reconciliation Twitter data" sounds like it's just about running some semi-automated process. This is a change where you actually reconciled some more people, so it would be better if the commit message reflected that. 

(Haven't looked at the actual change yet — that'll be for tomorrow morning)

tmtmtmtm commented:
@octopusinvitro when you say: 

```
GutierrezConcha https://twitter.com/GutierrezConcha --> Deputy
ConchaGutierrez https://twitter.com/ConchaGutierrez --> ?
In this case it was not clear so I left both.
```

Do you mean you left both _in_, or left both _out_? I'd assumed you meant that you had included them both, but it looks like you omitted them.

octopusinvitro commented:
@tmtmtmtm ?
Let me take a look.

tmtmtmtm commented:
I think we're getting too many low quality names here with things like")", "***Lourdes Ciuro ***", "Ascensióndelasheras", etc., so I think this is going to have to do the [Irish Shuffle](https://github.com/everypolitician/everypolitician-data/pull/13588) as well.

davewhiteland commented:
For a cleaner example of the _Irish shuffle_ (which is nudging the `names` column so it's not called anything the bot is going to automagically add to the data, but telling the reconciler to still use that column when comparing/matching/fuzzing against the existing names), see [this commit and its `temp_name` shuffle](https://github.com/everypolitician/everypolitician-data/pull/14399/commits/e5872d0e70eb7577bc1693493b8ede580000f07e)

octopusinvitro commented:
Special cases:

**5bfa6927-a3a8-4952-9fc7-32693a7e0dfa**

*  1147635685, Concha Gutierrez, [GutierrezConcha](https://twitter.com/GutierrezConcha)
*   337595761, Concha.gutirrez,  [ConchaGutierrez](https://twitter.com/ConchaGutierrez)

Kept both

**62b92498-42d3-485d-8b57-f88a9fa02b59**

*  129954202, Teófila Martínez, [teofilamartinez](https://twitter.com/teofilamartinez) --> Mayor
* 3130765065, Teófila Martínez, [teofila_es](https://twitter.com/teofila_es) --> Diputee

Deleted the first once

**fbd89aac-a9ac-4248-a0c7-c4bd80954e6f**

* 2921983427, Mar G. P., [margpuig](https://twitter.com/margpuig) --> María del Mar García Puig

Added manually to the reconciliation file, since she exists in our data and `twitter.html` even showed her to me on the right hand side, but for another politician, while this incoming one was showing other politicians to the right as well (?)

octopusinvitro commented:
Closing as this was redone using the Irish Shuffle in #16986

octopusinvitro commented:
Closing as this was redone using the Irish Shuffle in #16986
